src/posit/connect/packages.py

```diff
@@ -0,0 +1,231 @@
+from posit.connect.paginator import Paginator
+
+from .resources import Active, ActiveFinderMethods, ActiveSequence
+
+
+class ContentPackage(Active):
+    class _Package(TypedDict):
+        language: Required[str]
+        name: Required[str]
+        version: Required[str]
+        hash: Required[Optional[str]]
+
+    def __init__(self, ctx, /, **attributes: Unpack[_Package]):
+        # todo - passing "" is a hack since path isn't needed. Instead, this class should inherit from Resource, but ActiveSequence is designed to operate on Active. That should change.
+        super().__init__(ctx, "", **attributes)
+
+
+class ContentPackages(ActiveFinderMethods["ContentPackage"], ActiveSequence["ContentPackage"]):
+    """A collection of packages."""
+
+    def __init__(self, ctx, path):
+        super().__init__(ctx, path, "name")
+
+    def _create_instance(self, path, /, **attributes):  # noqa: ARG002
+        return ContentPackage(self._ctx, **attributes)
+
+    def fetch(self, **conditions):
+        try:
+            return super().fetch(**conditions)
+        except ClientError as e:
+            if e.http_status == 204:
+                return []
+            raise e
+
+    def find(self, uid):
+        raise NotImplementedError("The 'find' method is not support by the Packages API.")
+
+    class _FindBy(TypedDict, total=False):
+        language: NotRequired[Literal["python", "r"]]
+        """Programming language ecosystem, options are 'python' and 'r'"""
+
+        name: NotRequired[str]
+        """The package name"""
+
+        version: NotRequired[str]
+        """The package version"""
+
+        hash: NotRequired[Optional[str]]
+        """Package description hash for R packages."""
+
+    @overload
+    def find_by(self, **conditions: Unpack[_FindBy]):
+        """
+        Find the first record matching the specified conditions.
+
+        There is no implied ordering, so if order matters, you should specify it yourself.
+
+        Parameters
+        ----------
+        **conditions : Unpack[_FindBy]
+            Conditions for filtering packages. The following keys are accepted:
+
+        language : {"python", "r"}, not required
+            Programming language ecosystem, options are 'python' and 'r'
+
+        name : str, not required
+            The package name
+
+        version : str, not required
+            The package version
+
+        hash : str or None, optional, not required
+            Package description hash for R packages.
+
+        Returns
+        -------
+        Optional[T]
+            The first record matching the specified conditions, or `None` if no such record exists.
+        """
+
+    @overload
+    def find_by(self, **conditions): ...
+
+    def find_by(self, **conditions):
+        return super().find_by(**conditions)
+
+
+class ContentPackagesMixin(Active):
+    """Mixin class to add a packages attribute."""
+
+    @property
+    @requires(version="2024.10.0-dev")
+    def packages(self):
```

Comment:
Does this make the property disappear when the version is too low? 

Or does it present, but throws on access?

Reply:
It's available, but throws a NotImplementedError on access.

Reply:
Cool!